### DISCUSSION TITLE
Windows 10 IKEv2 Error: IKE Authentication credentials are unacceptable

### DISCUSSION THREAD

**Discussion by ww9rivers** [strongswan/strongswan#2575] [Q&A] [answered]
**System info**:

- **OS**: Ubuntu 24.04.1 LTS
- **Kernel version**: 6.8.0-51-generic
- **strongSwan** version(s): 5.9.13-2ubuntu4

I set up StrongSwan with Let's Encrypt certbot generated server cert, with user credentials in `/etc/ipsec.secrets` but I keep getting this error.

Attached is a copy of all the syslog messages for a IKEv2 session, with public IP address and hostname removed.
[syslog-strongswan.txt](https://github.com/user-attachments/files/18207131/syslog-strongswan.txt)

I am new to StrongSwan - I read a few posts here with the same error message but am unable to find a solution.

Happy holidays and thanks in advance for any help.

**Comment by Wonder2018** [comment]
### Notice

1. The ```extendedKeyUsage``` of the client certificate must includes ```clientAuth```
2. The ```extendedKeyUsage``` of the server certificate must includes ```serverAuth```
3. Server certificates **cannot** use wildcard domain certificates
4. ```connections>xxx>local>id``` must be exactly the same as the ```CN``` of the server certificate
5. The ```CN``` value of the certificate must be fully included in ```alt_names```. Because during ```EAP-TLS``` authentication, ```StrongSwan``` only recognizes ```alt_names```, but users will use ```CN``` for identification
6. When configuring VPN on Windows, you must use the ```CN``` of the server certificate to connect. If ```CN``` is an ```IP```, use ```IP```, and if ```CN``` is a ```domain name```, use ```domain name```. You can use a custom domain name and resolve it to the server's IP address in the ```hosts``` file


### Step

The following steps can be taken to generate client and server certificates:

## 1. create CA for server

```bash
openssl genpkey -algorithm RSA -out ca.key -pkeyopt rsa_keygen_bits:2048
openssl req -new -x509 -key ca.key -out ca.crt -days 3650 -subj "/CN=XXXXXXXXXXX"
```

## 2. create Server cert

> file server.cnf

```ini
# Client or server certificate request template,
# If the server certificate does not meet the following requirements, it will cause Windows login failure

[req]
distinguished_name = req_distinguished_name
x509_extensions = v3_req
prompt=no

[req_distinguished_name]
# CN must be fully included in alt_name, the username used by the user when logging in,
# The server certificate cannot use wildcard certificates
# set your domain
CN=wonderland.inc

[v3_req]
subjectAltName=@alt_names
basicConstraints=CA:FALSE
keyUsage=digitalSignature, keyEncipherment
# The client certificate must include 'contentAuth', and the server certificate must include 'serverless Auth'
extendedKeyUsage=serverAuth, clientAuth

[alt_names]
# If the generated client certificate CN must be fully included in alt_name,
# During EAP-TLS authentication, StrongSwan only recognizes alt_name, but CN is used for user identification
DNS.1=wonderland.inc
```

> gen server cert

```sh
# create server key
openssl genpkey -algorithm RSA -out server.key -pkeyopt rsa_keygen_bits:2048

# create CSR (Certificate Signing Request), Includes required SAN information and EKU expansion
openssl req -new -key server.key -subj "/CN=wonderland.inc/O=XXXXX/C=XX" -reqexts v3_req -out server.csr -config <(cat /etc/ssl/openssl.cnf server.cnf)

# Sign CSR with your CA and generate a certificate with the required extensions
openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -days 365 -out server.crt -sha256 -extensions v3_req -extfile server.cnf

```

## 3.  install server cert

    1. copy your ```ca.crt``` to ```/etc/swanctl/x509ca/ca.pem```
    2. copy your ```server.crt``` to ```/etc/swanctl/x509/server.pem```
    3. copy your ```server.key``` to ```server.key```

## 4. create Client cert

> file client.cnf

```ini
[req]
distinguished_name = req_distinguished_name
x509_extensions = v3_req
prompt=no

[req_distinguished_name]
C=XX
O=XXX
# This will become the username used for login
CN=client

[v3_req]
subjectAltName=@alt_names 
basicConstraints=CA:FALSE
keyUsage=digitalSignature, keyEncipherment
extendedKeyUsage=clientAuth,serverAuth

[ alt_names]
# This must be the same as CN
DNS.1=client
```

> gen client cert

```sh
# create clientkey
openssl genpkey -algorithm RSA -out server.key -pkeyopt rsa_keygen_bits:2048

# create CSR (Certificate Signing Request), Includes required SAN information and EKU expansion
openssl req -new -key server.key -subj "/CN=client/O=XXXXX/C=XX" -reqexts v3_req -out server.csr -config <(cat /etc/ssl/openssl.cnf server.cnf)

# Sign CSR with your CA and generate a certificate with the required extensions
openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -days 365 -out server.crt -sha256 -extensions v3_req -extfile server.cnf

# create pks file for windows
openssl pkcs12 -export -out client.pfx -inkey client.pem -in client.crt

```

## 5. install ca cert on windows

You need to import the ca.crt into the trusted CA area of the customer storage space, not the system storage space

## 6. install client cert on windows

    1. You need to double-click the client.pks file and select the customer storage space. 
    2. Then enter the password you set when creating the PKS file. 
    3. Check 'Enable strong private key protection'(The first checkbox on the password input interface) 
    4. Check 'Include all extended attributes'(The last checkbox on the password input interface). 
    5. Then click Next and set a password for your private key. 

> If you are using an early Windows 10 or earlier Windows system. You need to set the private key protection level to 'high'

**Reply by ww9rivers** [reply]
Thank you very much for responding!!

I'm trying to use a LetsEncrypt server cert instead of Strongswan `ipsec` generated CA + server certs so that I don't have deliver the CA cert to the clients. That should work (I got the idea from this script: https://github.com/jawj/IKEv2-setup/blob/master/setup.sh).

Inspecting the Letscrypt server cert, I think it meets the requirements you mentioned:
2. The `extendedKeyUsage` of the server certificate must includes `serverAuth`: the cert has this:
```
        X509v3 extensions:
            X509v3 Extended Key Usage:
                TLS Web Server Authentication, TLS Web Client Authentication
```
Also, the cert has the server's DNS name in its `Subject: CN =` field and the "X509v3 Subject Alternative Name" extension. The server cert is definitely not a wildcard cert.

As for the client, again I did not create a client cert. as Windows does support "User name and password" for sign-in. Is client cert authentication a must for Strongswan?

**Edit**: One more question - Does the "IKE Authentication credentials are unacceptable" message mean the client's credentials or something else? I understand that may be a Windows message so it may not be known from Strongswan's perspective.

**Reply by Wonder2018** [reply]
you should check this:

copy or make a link --> ```/etc/letsencrypt/live/${YOUR_DOMAIN}/chain.pem``` to ```/etc/swanctl/x509ca/vpn-ca.pem```
copy or make a link --> ```/etc/letsencrypt/live/${YOUR_DOMAIN}/fullchain.pem``` to ```/etc/swanctl/x509/server.pem```
copy or make a link --> ```/etc/letsencrypt/live/${YOUR_DOMAIN}/privkey.pem``` to ```/etc/swanctl/private/server.key```

**Reply by Wonder2018** [reply, marked as answer]
> you should check this:
> 
> copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/chain.pem` to `/etc/swanctl/x509ca/vpn-ca.pem` copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/fullchain.pem` to `/etc/swanctl/x509/server.pem` copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/privkey.pem` to `/etc/swanctl/private/server.key`



I may find the cause of the problem

At the end of the few lines in file [syslog-strongswan.txt](https://github.com/user-attachments/files/18207131/syslog-strongswan.txt) that you provided 

```log
2024-12-20T05:23:52.918637+00:00 pine charon: 11[IKE] no private key found for '{{vpn-hostname}}'
2024-12-20T05:23:52.918759+00:00 pine charon: 11[ENC] generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]
2024-12-20T05:23:52.918883+00:00 pine charon: 11[NET] sending packet: from 10.1.10.3[4500] to {{client-ip}}[4500] (80 bytes)
2024-12-20T05:23:52.919022+00:00 pine charon: 11[IKE] IKE_SA ikev2-eap-mschapv2[2] state change: CONNECTING => DESTROYING
```

strongswan can't find your private key of your server cert. 

You need to check if these files have been placed in Strongswan's certificate directory, as I previously replied

**Reply by ww9rivers** [reply]
> you should check this:
> 
> copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/chain.pem` to `/etc/swanctl/x509ca/vpn-ca.pem`
> copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/fullchain.pem` to `/etc/swanctl/x509/server.pem`
> copy or make a link --> `/etc/letsencrypt/live/${YOUR_DOMAIN}/privkey.pem` to `/etc/swanctl/private/server.key`

I am guessing this might have been the answer.

I did those, but was still unable to get things to work. Then I used the [IKEv2-setup](https://github.com/jawj/IKEv2-setup/) on my VPN server and I was able to connect - even though I remember [someone else had an issue](https://github.com/jawj/IKEv2-setup/issues/163) with the script.

Thank you for all the help.

**Comment by ww9rivers** [comment]
Inspecting the syslog-strongswan.txt file, I found these lines that look suspicious:
```
2024-12-20T05:23:52.810407+00:00 pine charon: 09[IKE] sending cert request for "CN=VPN CA"
 . . . . .
2024-12-20T05:23:52.909027+00:00 pine ipsec[10802]: 11[IKE] no private key found for '{{vpn-hostname}}'
2024-12-20T05:23:52.909176+00:00 pine charon: 11[IKE] received cert request for unknown ca with keyid ab:30:d3:af:4b:d8:f1:6b:58:69:ee:45:69:29:da:84:b8:73:94:88
2024-12-20T05:23:52.909368+00:00 pine ipsec[10802]: 11[ENC] generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]
```
Why is it asking for `CN=VPN CA`? I am not able to fine that configured anywhere.

This is my `/etc/ipsec.conf` file:
```
config setup
    charondebug="ike 2, knl 2, cfg 2"
    uniqueids=no

conn ikev2-eap-mschapv2
    auto=add
    ike=aes256-sha256-modp1024!
    esp=aes256-sha256!
    dpdaction=clear
    dpddelay=300s
    rekey=no
    left=%any
    leftid=@{{vpn-hostname}}
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
    right=%any
    rightdns=10.1.10.1, 8.8.4.4
    rightsourceip=10.10.10.0/24
    rightauth=eap-mschapv2
    eap_identity=%identity
```
